content/blog/2020-11-04-unified-theory-gc/index.md

```diff
@@ -0,0 +1,176 @@
+not have long pauses during execution, but rather pays for its overhead 
+incrementally. Reference counting's major drawback is the difficulty presented
+by cycles appearing the link structure, which prevent objects' reference counts
+from ever reaching 0.
+
+## Highlights
+
+Bacon et al. begin by presenting the two fundamental garbage collection
+algorithms, reference counting and tracing, as dual views of the same underlying
+structure. This should hit you a bit like hearing that butter and I Can't
+Believe It's Not Butter have the same deep structure.
+
+The deep comparison they make between the two paradigms is developed in a number
+of ways:
+
+* **Duality of operations:** Tracing traverses the live
+  objects on the heap in a technically analogous way to how reference
+  counting traverses dead objects. 
+
+* **Algorithm structure:** They make this more concrete by giving pseudocode algorithms
+  for each that really are strikingly similar.
+
+* **Fixed point forumation:** They give a formal presentation of garbage
+  collection from an abstract standpoint. In this abstract view, the goal of garbage
+  collection is to compute a fixed-point of a system of references where nodes'
+  reference counts accurately reflect in-edges from either (i) members of the
+  root set, or (ii) objects whose reference count is nonzero. The fixed-point
+  here could be imagined as the result of some iterate-to-convergence approach.
+
+They go on to demonstrate that a whole continuum of hybrid approaches between
+the two not only make sense in principle but in fact really are already in use
+in practice. Here is a list of the the garbage collection strategies in *rough*
+order from most mark-n-sweepy to most-reference-county:
+
+* **T: Tracing.** This is the classic tracing algorithm. The algorithm must
+  traverse the link graph of the heap to 'mark' objects live. The authors point
+  out that tail recursion can reduce the space cost of the traversal when singly
+  linked structures are present.
+
+* **GT: Tracing Generational Collection.** This form of collection uses a
+  "nursery" to improve the time spent on collection. The idea is that recently
+  allocated objects are more likely to be freed earlier, so we can spend our
+  garbage-collection time more efficiently by looking at newer objects more
+  often.
+
+* **Older First.** (Sketched only). In this approach, the oldest generation
+  is traced first instead of the nursery.
+
+* **GCH: Generational with Counted Heap.** Run normal tracing of the nursery,
+  but do reference counting on the mature remaining space of the heap.
+
+* **Train Algorithm.** (Presented but not analyzed). While this approach was
+  definitely interesting, and they drive home that it fits into their hybrid
+  model (car ordering provides implicit reference counts), I must admit I was not
+  able to glean much intuition about why the approach would be worth
+  implementing.
```

Comment:
Maybe this bullet could use one more sentence about what the approach actually is before adding this additional commentary?